### PR TITLE
#492 - builder for DataFetchingEnvironmentImpl

### PR DESCRIPTION
**Pull request by bbakerman**
#492 - builder for DataFetchingEnvironmentImpl

### CODE REVIEW

**Comment by bsideup** [comment]
@bbakerman could you please provide `.transform()` method? (see https://github.com/graphql-java/graphql-java/pull/479 for the ref)

**Comment by bbakerman** [comment]
@bsideup - where would you put the transform() method

on DataFetchingEvironment or the DataFetchingEvironmentImpl?

I thought about this and it logically belongs on the interface BUT the concrete builder is on the DataFetchingEvironmentImpl class.

Where would you put the transform so its meaningful?

I kind of did the equivalent via this

```
            DataFetchingEnvironment singleEnv = DataFetchingEnvironmentImpl.newDataFetchingEnvironment(environment)
                    .source(source)
                    .build();
```

**Comment by bbakerman** [comment]
I guess we could put it on DataFetchingEnvironmentImpl but that would require casting first which is weird

     DataFetchingEnvironment env
     DataFetchingEnvironment newEnv = ((DataFetchingEnvironmentImpl) env).transform( bld -> bld.source(newSource))

if we put it on the interface then the builder (from the impl) will become part of the Interface

```
@PublicApi
public interface DataFetchingEnvironment {

    public static DataFetchingEnvironmentImpl.Builder newBuilder() {
        ...
    }
```

Thats also weird.

I guess I could make Builder part of the interface itself

File: src/main/java/graphql/schema/DataFetchingEnvironment.java [hunk 1]
```diff
@@ -100,4 +108,18 @@
...
+
+
+    /**
+     * This allows you to transform this DataFetchingEnvironment object into another one.
+     *
+     * @param action the action to which a {@link DataFetchingEnvironmentBuilder} is passed
+     *
+     * @return a new transformed DataFetchingEnvironment
+     */
+    default DataFetchingEnvironment transform(Consumer<DataFetchingEnvironmentBuilder> action) {
```
Comment (andimarek) [inline src/main/java/graphql/schema/DataFetchingEnvironment.java]: I am not really happy with adding that to the interface to be honest.

Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingEnvironment.java]: Where would you put it? On the DFEImpl instead?  

I can imagine DataFetchers that wrap other data fetchers might want to use this to "tweak" the environment before calling the wrapped DF.

But its just sugar on `Builder.newDataFetchingEvironment(currentEv).build() so I can remove it if you want.  I am not wedded to it in this case

Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingEnvironment.java]: IN fact here is one from the graphql base code

````
public class UnbatchedDataFetcher implements BatchedDataFetcher {

    private final DataFetcher delegate;

    public UnbatchedDataFetcher(DataFetcher delegate) {
        this.delegate = delegate;
    }


    @Override
    public Object get(DataFetchingEnvironment environment) {
        List<Object> sources = environment.getSource();
        List<Object> results = new ArrayList<>();
        for (Object source : sources) {

            DataFetchingEnvironment singleEnv = environment
                    .transform(builder -> builder.source(source));
            results.add(delegate.get(singleEnv));
        }
        return results;
    }
}
````

Comment (bbakerman) [inline src/main/java/graphql/schema/DataFetchingEnvironment.java]: FYI I have removed it

**Pull request merged**